### PR TITLE
Article - Node vs Python APIs

### PR DESCRIPTION
**Pull request by cyberShaw**
Added article on Node and Python APIs.

Provided a comparison between the two technologies and how they would be suitable for writing APIs.
also, provided some code snippets and explained them. Please review and merge 😄

### CODE REVIEW

File: articles/node-vs-python/index.md [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+- [Node vs Python: The Differences](#node-vs-python:-the-differences)
+- [APIs in Node](#apis-in-node)
+- [APIs in Python](#apis-in-python)
+- [Further Reading](#further-reading)
+
+### The History
+
+#### Node
+
+Node is a server-side javascript runtime. JavaScript was originally designed to run within a browser, inside a JavaScript engine like V8. To expand the capabilities of JavaScript outside the browser, Node was developed in **2009** by **Ryan Dahl**. Node combined the capabilites of the V8 JavaScript engine, an event loop and low level calls into a robust server-side scripting runtime. It soon grew into the preferred language by web developers.
```
Comment (hectorkambow) [inline articles/node-vs-python/index.md:33]: @cyberShaw  good morning Harsha  - just starting a preliminary review on your article - wanted to ask if this 1st sentence here is complete - Im not sure. 🤔 
Node is a server side JavaScript runtime ???

Comment (cyberShaw) [reply articles/node-vs-python/index.md]: Yes haha. Actually the definition of node was taken directly from the Node JS docs.

"Node is a Javascript Runtime based on the V8 Engine" 

So I think "Node is a Javascript Runtime" could also be written.

**Comment by hectorkambow** [comment]
Good afternoon @mhoulmiere 👋  Wanted to ask if you could give this article a tech review before we published 🚀 
Thank you

File: articles/node-vs-python/index.md [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+description: Node.js and Python are among the most popular languages used to write APIs and backend services. Let's look at a comparison between the two most popular languages out there.
+author: saiharsha-balasubramaniam
+date: 2020-08-13T00:00:00-17:00
+topics: [Node.js]
+excerpt_separator: <!--more-->
+images:
+  - url: /engineering-education/node-vs-python/hero.jpg
+    alt: node.js vs Python
+---
+JavaScript and Python are the most popular languages among web developers. The idea of separating the front-end from the back-end as much as possible has birthed the idea of the JAMStack. JAMStack stands for **JavaScript, APIs and Markup**. It describes the idea of a website being completely static and client-side rendered. The website uses API calls to load all dynamic content. Let us look at the two most popular languages for writing APIs, Node and Python to compare and contrast the two.
```
Comment (mhoulmiere) [inline articles/node-vs-python/index.md:16]: All references to node.js should be written as node.js not "Node"

Nitpicky and not something you have to change, but I would hedge the statement in the first sentence to something like "JavaScript and Python are two of the most popular languages..."

File: articles/node-vs-python/index.md [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+<!--more-->
+
+### Table of Contents
+
+- [The History](#the-history)
+- [Node vs Python: The Differences](#node-vs-python:-the-differences)
+- [APIs in Node](#apis-in-node)
+- [APIs in Python](#apis-in-python)
+- [Further Reading](#further-reading)
```
Comment (mhoulmiere) [inline articles/node-vs-python/index.md:25]: Nice! Every article should have this!

File: articles/node-vs-python/index.md [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+- [Type of Projects](#type-of-projects)
+- [Frameworks and Expandability](#frameworks-and-expandability)
+- [Package Management](#package-management)
+
+#### Performance and Architecture
+Python and Node.js are interpreted languages. Therefore, they are comparatively slower than compiled languages like C++.
+
+Python is a single threaded and single flowing language. Though it is optimized by frameworks such as Django, it is slow. Python runs well on servers, but is quite unsuitable for mobile. The limit in processing power is a bottleneck to Python's performance.
+
+Node.js has the upper hand here. Node.js is heavily optimized by Google's V8 JavaScript Engine, making it performant. Node.js follows an event loop based architecture, which makes it non-blocking. Asynchronous requests can be handled. This means that the runtime can execute multiple requests without getting blocked.
```
Comment (mhoulmiere) [inline articles/node-vs-python/index.md:47]: "Asynchronous requests are supported and this means..." is a little bit more correct (and sounds better IMO).

File: articles/node-vs-python/index.md [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+#### Syntax and Ease of Use
+Python has extremely readable syntax. This makes it the preferred programming language for beginners to work on. It omits the use of curly braces and uses indentation to define scope.
+
+Node.js follows the JavaScript syntax, which is harder to understand than Python. But for programmers that are familiar with C/C++, Node.js syntax is similar and poses no challenge.
+
+#### Type of Projects
+Both Python and Node.js are used for web development and full-stack projects.
```
Comment (mhoulmiere) [inline articles/node-vs-python/index.md:55]: I would expound on this (web development vs full-stack projects) by and give examples.

File: articles/node-vs-python/index.md [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+Python is heavily used as a part of Data Science and Machine Learning projects. It has libraries like numpy, pandas, scikit, opencv and tensorflow that make it perfect for:
+
+- Neural Networks
+- Face Recognition
+- Data Processing
+
+Python is also preferred for small, lightweight websites. Flask, the micro-framework and Jinja, the templating engine is used by web developers.
```
Comment (mhoulmiere) [inline articles/node-vs-python/index.md:63]: You should give examples/expound on this and the second sentence is awkward/hard to follow.

File: articles/node-vs-python/index.md [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+Due to the event loop architecture of Node.js, it is used for websites that involve a lot of concurrent processes. It is also suitable for Internet of Things solutions and projects.
```
Comment (mhoulmiere) [inline articles/node-vs-python/index.md:65]: Give examples and expound on "websites that involve a lot fo concurrent processes."

File: articles/node-vs-python/index.md [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+
+#### Frameworks and Expandability
+Python uses an interpreter called CPython that improves the performance of Python. There are many frameworks and libraries for full stack development.
+- [Django](https://www.djangoproject.com/) - A python web framework based on the Model-View-Controller Architecture
+- [Flask](https://flask.palletsprojects.com/en/1.1.x/) - A python micro-framework for web development
+- [Tensorflow](https://www.tensorflow.org/) - A python based Machine Learning platform
```
Comment (mhoulmiere) [inline articles/node-vs-python/index.md:71]: There are more interpreters than just CPython. Also, I would expound on the pros/cons of each of the frameworks. Also, you should think about adding a blurb on what a micro-framework is vs a framework.

File: articles/node-vs-python/index.md [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+
+Node.js has a ton of libraries that make it extensible, scalable and feature-rich.
+- [Express](https://expressjs.com/) - Unopinionated web framework for Node.js.
+- [Adonis](https://adonisjs.com/) - Full-Stack MVC Framework.
+- [Koa](https://koajs.com/) - Robust Foundation for Web Applications.
```
Comment (mhoulmiere) [inline articles/node-vs-python/index.md:76]: You should expound on the pros/cons of each of the frameworks.

File: articles/node-vs-python/index.md [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+
+#### Package Management
```
Comment (mhoulmiere) [inline articles/node-vs-python/index.md:78]: You should provide details on what a package manager is exactly and what benefit it provides when using the language.

File: articles/node-vs-python/index.md [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+
+- Neural Networks
+- Face Recognition
+- Data Processing
+
+Python is also preferred for small, lightweight websites. Flask, the micro-framework and Jinja, the templating engine is used by web developers.
+
+Due to the event loop architecture of Node.js, it is used for websites that involve a lot of concurrent processes. It is also suitable for Internet of Things solutions and projects.
+
+#### Frameworks and Expandability
```
Comment (mhoulmiere) [inline articles/node-vs-python/index.md:67]: You should provide details on what frameworks are and what benefits they provide.

File: articles/node-vs-python/index.md [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+
+To keep track of packages in a project, Node.js uses a file called **package.json**
+
+To install a Node.js package, just type:
+
+```
+npm i <package-name>
+```
+
+### APIs in Node.js
```
Comment (mhoulmiere) [inline articles/node-vs-python/index.md:99]: You should provide a brief explanation of what an API is and why you would want to build one. Also, you should provide your opinion (and stated as an opinion) on which of the two you would use and why (pros and cons of each) somewhere in this section.

File: articles/node-vs-python/index.md [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+
+app.get("/read", (req, res) => {
+  connection.query("SELECT * FROM sample", (err, res, fields) => {
+    if (err) throw err;
+    console.log(res);
+  });
+});
+```
+
+In the snippet above, we initialize a MySQL connection using the Node.js MySQL library. Whenever a get request is sent to the _/read_ endpoint, the database is read.
```
Comment (mhoulmiere) [inline articles/node-vs-python/index.md:146]: Your explanation of the code is lacking. You should provide comments in the code that better explain what the various lines of code are for.

File: articles/node-vs-python/index.md [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+
+To install a Node.js package, just type:
+
+```
+npm i <package-name>
+```
+
+### APIs in Node.js
+
+Let us dive into creating an API in Node.js our API has an endpoint _/read_ which reads data from a table, _sample_ in a MySQL Database.
```
Comment (mhoulmiere) [inline articles/node-vs-python/index.md:101]: You should expound on MySQL and provide a link to an educational article on it. Anyone reading this article that actually wants to do this tutorial will be unable to because you have not given them all of the needed pieces.

File: articles/node-vs-python/index.md [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+    if (err) throw err;
+    console.log(res);
+  });
+});
+```
+
+In the snippet above, we initialize a MySQL connection using the Node.js MySQL library. Whenever a get request is sent to the _/read_ endpoint, the database is read.
+
+### APIs in Python
+Let us build a similar API that reads a database (SQLAlchemy) using Python.
```
Comment (mhoulmiere) [inline articles/node-vs-python/index.md:149]: You should expound on SQLAlchemy and provide a link to an educational article on it. Anyone reading this article that actually wants to do this tutorial will be unable to because you have not given them all of the needed pieces.

File: articles/node-vs-python/index.md [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+        query = conn.execute("select * from sample") # Query and return json result
+        return jsonify(query)
+
+api.add_resource(Read, '/read')
+
+if __name__ == '__main__':
+     app.run(port='5002')
+```
+
+In the snippet above, we initialize a MySQL connection using the SQLAlchemy library in Python. The Flask library is used for creating API endpoints.
```
Comment (mhoulmiere) [inline articles/node-vs-python/index.md:174]: Your explanation of the code is lacking. You should provide comments in the code that better explain what the various lines of code are for.

File: articles/node-vs-python/index.md [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+    alt: node.js vs Python
+---
+JavaScript and Python are the most popular languages among web developers. The idea of separating the front-end from the back-end as much as possible has birthed the idea of the JAMStack. JAMStack stands for **JavaScript, APIs and Markup**. It describes the idea of a website being completely static and client-side rendered. The website uses API calls to load all dynamic content. Let us look at the two most popular languages for writing APIs, Node and Python to compare and contrast the two.
+<!--more-->
+
+### Table of Contents
+
+- [The History](#the-history)
+- [Node vs Python: The Differences](#node-vs-python:-the-differences)
+- [APIs in Node](#apis-in-node)
```
Comment (mhoulmiere) [inline articles/node-vs-python/index.md:23]: These two links no longer match the headings so the links are broken.

File: articles/node-vs-python/index.md (hunk 4 above)
Comment (cyberShaw) [reply articles/node-vs-python/index.md:47]: Sure, shall make the suggested change! 👍

File: articles/node-vs-python/index.md (hunk 17 above)
Comment (cyberShaw) [reply articles/node-vs-python/index.md:23]: Thanks a ton for suggesting these changes. I'll work on them and commit my changes!

**Comment by cyberShaw** [comment]
@hectorkambow I've made the changes suggested in the review!
Please check it out 😄
